serializers/src/main/java/io/pravega/schemaregistry/serializers/AbstractSerializer.java

```diff
@@ -0,0 +1,114 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.schemaregistry.serializers;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.util.BitConverter;
+import io.pravega.schemaregistry.client.SchemaRegistryClient;
+import io.pravega.schemaregistry.codec.Codec;
+import io.pravega.schemaregistry.contract.data.EncodingId;
+import io.pravega.schemaregistry.contract.data.SchemaInfo;
+import io.pravega.schemaregistry.contract.data.VersionInfo;
+import io.pravega.schemaregistry.schemas.Schema;
+import lombok.Getter;
+import lombok.SneakyThrows;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicReference;
+
+abstract class AbstractSerializer<T> extends BaseSerializer<T> {
+    private static final byte PROTOCOL = 0x1;
+    private static final int HEADER_LENGTH = Byte.BYTES + Integer.BYTES;
+
+    private final String groupId;
+    
+    private final SchemaInfo schemaInfo;
+    private final AtomicReference<EncodingId> encodingId;
+    private final boolean encodeHeader;
+    private final SchemaRegistryClient client;
+    @Getter
+    private final Codec codec;
+    private final boolean registerSchema;
+    
+    protected AbstractSerializer(String groupId,
+                                 SchemaRegistryClient client,
+                                 Schema<T> schema,
+                                 Codec codec,
+                                 boolean registerSchema, 
+                                 boolean encodeHeader) {
+        Preconditions.checkNotNull(groupId);
+        Preconditions.checkNotNull(client);
+        Preconditions.checkNotNull(codec);
+        Preconditions.checkNotNull(schema);
+        
+        this.groupId = groupId;
+        this.client = client;
+        this.schemaInfo = schema.getSchemaInfo();
+        this.registerSchema = registerSchema;
+        this.encodingId = new AtomicReference<>();
+        this.codec = codec;
+        this.encodeHeader = encodeHeader;
+        initialize();
+    }
+    
+    private void initialize() {
+        VersionInfo version;
+        if (registerSchema) {
+            // register schema
+            version = client.addSchema(groupId, schemaInfo);
+        } else {
+            // get already registered schema version. If schema is not registered, this will throw an exception. 
+            version = client.getVersionForSchema(groupId, schemaInfo);
+        }
+        if (encodeHeader) {
+            encodingId.set(client.getEncodingId(groupId, version, codec.getCodecType().getName()));
+        }
+    }
+    
+    @SneakyThrows(IOException.class)
+    @Override
+    public ByteBuffer serialize(T obj) {
+        ByteArrayOutputStream dataStream = new ByteArrayOutputStream();
```

Comment:
This uses a 32 byte array and re-copies all the data so far to a new array twice as big every time it goes over the size limit.
If we could have a better estimate of the size we could avoid some of these copies. Is there a way to do that? 
If not perhaps we should just pass 128 or something to the constructor to make it a little better.

Reply:
hmm.. we dont have a way to estimate the size of serialized blob, neither do we have a way to estimate the size of encoded blob. 
I can build a SegmentedByteArrayOutputStream that does not have a single array backing the data but a linked list of small arrays. It creates new array of fixed size each time the size spills over.
When eventually stream.toByteArray is called, it knows the size and can create a byte array and copy all of the data. 

That way we avoid all these copies for an expanding array. But downside compared to EnhancedByteArrayOutputStream is, it wont expose its backing array to the caller. 

I think thats ok, and perhaps better than 7 copies for a 1kb serialized data.

Reply:
thinking some more, since a list would itself induce additional memory footprint than the byte buffer copy we are trying to save on, esp if we have small buffers. 

maybe initializing with a larger buffer is better choice. but lets evaluate the pros and cons:

larger allotment vs list
con
1. larger allotment even for smaller data
pro
1. less copies
lets say we start with 128 bytes instead of 32 bytes. that will save on a copy of 32 bytes and 64 bytes before we hit 128 bytes after which its the same. 
So what we save here is copying over 96 bytes. 

but with 1kb, the copies are 32 + 64 + 128 + 256 + 512 = 992 extra bytes. 

array list (linked lists would induce memory per element.. array lists are better for high number of elements/smaller buffers):
with 1 kb, at 32 byte buffers, we will get 32 buffers, which means 3 expansions on array list, maybe around 300 extra bytes for array list including the references to the buffers. The number of references can be reduced with larger buffers. 

maybe larger buffers being copied over is bad enough, and to avoid that expansion we can perhaps use an array list.. but then we lose the benefit of using the output stream's backing array implicitly. 

Maybe just using a higher buffer with enhancedbytearrayoutputstream is a good start. we can improve on it subsequently with more experimentation.

Reply:
Update: I have used EnhancedByteArrayOutputStream which saves on the final copy. 
It did not have a constructor to set the buffer to a higher size. So i have not made that change yet. 
But 128 bytes and 32 bytes initialization will have a 96 byte copy phase as additional overhead. 
I will explore all options and i have created an issue for the same. 
https://github.com/pravega/schema-registry/issues/52

Reply:
It's worth noting that copy cost is not linear in number of bytes below 1kb. 
(There are hardware instructions to copy 1, 4, 8, 16, 32, and 64 bytes at a time on modern X86), plus loop unrolling and instruction parallelism, alignment, etc) This means there is significant constant factor overhead at small sizes compared to the high throughput copies of large contiguous arrays.

I agree that ArrayList of arrays is much better than linked list of arrays.
If we want to go the list of arrays approach, I would recommend that each element be at least 1kb.

I think there is an opportunity to create a better abstraction for this sort of thing that could easily go into common. I don't think we need to do it in this PR though.

Reply:
have created issue https://github.com/pravega/schema-registry/issues/52 to explore this outside of this PR.